File: tensorflow/core/common_runtime/eager/eager_operation_test.cc
/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/common_runtime/eager/eager_operation.h"

#include "tensorflow/core/lib/core/status_test_util.h"
#include "tensorflow/core/platform/errors.h"
#include "tensorflow/core/platform/test.h"

namespace tensorflow {
namespace {

TEST(EagerOperationTest, DeviceName) {
  StaticDeviceMgr device_mgr(DeviceFactory::NewDevice(
      "CPU", {}, "/job:localhost/replica:0/task:0/device:CPU:0"));
  auto ctx = new EagerContext(
      SessionOptions(),
      tensorflow::ContextDevicePlacementPolicy::DEVICE_PLACEMENT_SILENT, false,
      &device_mgr, false, nullptr, nullptr, nullptr,
      /*run_eager_op_as_function=*/true);

  auto op = new EagerOperation(ctx);

  TF_ASSERT_OK(op->SetDeviceName("/device:DONTHAVE"));
  EXPECT_EQ("/device:DONTHAVE:*", op->DeviceName());

  TF_ASSERT_OK(op->SetDeviceName(""));
  EXPECT_EQ("", op->DeviceName());

  TF_ASSERT_OK(op->SetDeviceName("/job:localhost"));
  EXPECT_EQ("/job:localhost", op->DeviceName());

  EXPECT_NE(absl::OkStatus(), op->SetDeviceName("/not/a/valid/name"));

  delete op;
  ctx->Unref();
}

TEST(EagerOperationTest, EagerFunctionParamsAndStepId) {
  StaticDeviceMgr device_mgr(DeviceFactory::NewDevice(
      "CPU", {}, "/job:localhost/replica:0/task:0/device:CPU:0"));
  auto ctx = new EagerContext(
      SessionOptions(),
      tensorflow::ContextDevicePlacementPolicy::DEVICE_PLACEMENT_SILENT, false,
      &device_mgr, false, nullptr, nullptr, nullptr,
      /*run_eager_op_as_function=*/true);

  tensorflow::FunctionDef function_def;
  CHECK(tensorflow::protobuf::TextFormat::ParseFromString(
      "    signature {"
      "      name: 'DummyFunction'"
      "    }",
      &function_def));
  TF_ASSERT_OK(ctx->AddFunctionDef(function_def));

  auto op = new EagerOperation(ctx);
  EXPECT_FALSE(op->eager_func_params().has_value());

  string device_name = "/job:localhost/replica:0/task:0/device:CPU:0";
  TF_ASSERT_OK(op->SetDeviceName(device_name.c_str()));
  TF_ASSERT_OK(op->Reset("DummyFunction", device_name.c_str()));

  op->SetStepId(255);
  EXPECT_EQ(op->eager_func_params()->step_id.value(), 255);

  delete op;
  ctx->Unref();
}

}  // namespace
}  // namespace tensorflow

File: tensorflow/java/src/test/java/org/tensorflow/EagerOperationTest.java
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package org.tensorflow;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Unit tests for {@link EagerOperation} class. */
@RunWith(JUnit4.class)
public class EagerOperationTest {

  @Test
  public void failToCreateIfSessionIsClosed() {
    EagerSession session = EagerSession.create();
    session.close();
    try {
      new EagerOperation(session, 1L, new long[] {1L}, "Add", "add");
      fail();
    } catch (IllegalStateException e) {
      // expected
    }
  }

  @Test
  public void outputDataTypeAndShape() {
    try (EagerSession session = EagerSession.create();
        Tensor<Integer> t = Tensors.create(new int[2][3])) {
      EagerOperation op =
          opBuilder(session, "Const", "OutputAttrs")
              .setAttr("dtype", DataType.INT32)
              .setAttr("value", t)
              .build();
      assertEquals(DataType.INT32, op.dtype(0));
      assertEquals(2, op.shape(0)[0]);
      assertEquals(3, op.shape(0)[1]);
    }
  }

  @Test
  public void outputTensor() {
    try (EagerSession session = EagerSession.create()) {
      EagerOperation add =
          opBuilder(session, "Add", "CompareResult")
              .addInput(TestUtil.constant(session, "Const1", 2))
              .addInput(TestUtil.constant(session, "Const2", 4))
              .build();
      assertEquals(6, add.tensor(0).intValue());

      // Validate that we retrieve the right shape and datatype from the tensor
      // that has been resolved
      assertEquals(0, add.shape(0).length);
      assertEquals(DataType.INT32, add.dtype(0));
    }
  }

  @Test
  public void inputAndOutputListLengths() {
    try (EagerSession session = EagerSession.create()) {
      Output<Float> c1 = TestUtil.constant(session, "Const1", new float[] {1f, 2f});
      Output<Float> c2 = TestUtil.constant(session, "Const2", new float[] {3f, 4f});

      EagerOperation acc =
          opBuilder(session, "AddN", "InputListLength")
              .addInputList(new Output<?>[] {c1, c2})
              .build();
      assertEquals(2, acc.inputListLength("inputs"));
      assertEquals(1, acc.outputListLength("sum"));

      EagerOperation split =
          opBuilder(session, "Split", "OutputListLength")
              .addInput(TestUtil.constant(session, "Axis", 0))
              .addInput(c1)
              .setAttr("num_split", 2)
              .build();
      assertEquals(1, split.inputListLength("split_dim"));
      assertEquals(2, split.outputListLength("output"));

      try {
        split.inputListLength("no_such_input");
        fail();
      } catch (IllegalArgumentException e) {
        // expected
      }

      try {
        split.outputListLength("no_such_output");
        fail();
      } catch (IllegalArgumentException e) {
        // expected
      }
    }
  }

  @Test
  public void numOutputs() {
    try (EagerSession session = EagerSession.create()) {
      EagerOperation op =
          opBuilder(session, "UniqueWithCountsV2", "unq")
              .addInput(TestUtil.constant(session, "Const1", new int[] {1, 2, 1}))
              .addInput(TestUtil.constant(session, "Axis", new int[] {0}))
              .setAttr("out_idx", DataType.INT32)
              .build();
      assertEquals(3, op.numOutputs());
    }
  }

  @Test
  public void opNotAccessibleIfSessionIsClosed() {
    EagerSession session = EagerSession.create();
    EagerOperation add =
        opBuilder(session, "Add", "SessionClosed")
            .addInput(TestUtil.constant(session, "Const1", 2))
            .addInput(TestUtil.constant(session, "Const2", 4))
            .build();
    assertEquals(1, add.outputListLength("z"));
    session.close();
    try {
      add.outputListLength("z");
      fail();
    } catch (IllegalStateException e) {
      // expected
    }
  }

  @Test
  public void outputIndexOutOfBounds() {
    try (EagerSession session = EagerSession.create()) {
      EagerOperation add =
          opBuilder(session, "Add", "OutOfRange")
              .addInput(TestUtil.constant(session, "Const1", 2))
              .addInput(TestUtil.constant(session, "Const2", 4))
              .build();
      try {
        add.getUnsafeNativeHandle(1);
        fail();
      } catch (IndexOutOfBoundsException e) {
        // expected
      }
      try {
        add.shape(1);
        fail();
      } catch (IndexOutOfBoundsException e) {
        // expected
      }
      try {
        add.dtype(1);
        fail();
      } catch (IndexOutOfBoundsException e) {
        // expected
      }
      try {
        add.tensor(1);
        fail();
      } catch (IndexOutOfBoundsException e) {
        // expected
      }
    }
  }

  private static EagerOperationBuilder opBuilder(EagerSession session, String type, String name) {
    return new EagerOperationBuilder(session, type, name);
  }
}
